src/server/game/Handlers/MovementHandler.cpp

```diff
@@ -450,78 +451,107 @@ void WorldSession::HandleForceSpeedChangeAck(WorldPacket &recvData)
         default:
-            TC_LOG_ERROR("network", "WorldSession::HandleForceSpeedChangeAck: Unknown move type opcode: %u", recvData.GetOpcode());
+            TC_LOG_INFO("cheat", "WorldSession::HandleForceSpeedChangeAck: Player %s from account id %u kicked for incorrect data returned in an ack",
+                _player->GetName().c_str(), _player->GetSession()->GetAccountId());
+            _player->GetSession()->KickPlayer("incorrect data returned in an ack");
             return;
     }
 
-    // skip all forced speed changes except last and unexpected
-    // in run/mounted case used one ACK and it must be skipped.m_forced_speed_changes[MOVE_RUN} store both.
-    if (mover->m_forced_speed_changes[force_move_type] > 0)
+    if (pendingChange.movementCounter != movementCounter || std::fabs(speedSent - speedReceived) > 0.01f || moveTypeSent!= move_type)
     {
-        --mover->m_forced_speed_changes[force_move_type];
-        if (mover->m_forced_speed_changes[force_move_type] > 0)
-            return;
+        TC_LOG_INFO("cheat", "WorldSession::HandleForceSpeedChangeAck: Player %s from account id %u kicked for incorrect data returned in an ack",
+            _player->GetName().c_str(), _player->GetSession()->GetAccountId());
+        _player->GetSession()->KickPlayer("incorrect data returned in an ack");
+        return;
     }
 
-    if (!mover->GetTransport() && std::fabs(mover->GetSpeed(move_type) - newspeed) > 0.01f)
+    /* the client data has been verified. let's do the actual change now */
+    int64 movementTime = (int64)movementInfo.time + _timeSyncClockDelta;
+    if (_timeSyncClockDelta == 0 || movementTime < 0 || movementTime > 0xFFFFFFFF)
     {
-        if (mover->GetSpeed(move_type) > newspeed)         // must be greater - just correct
-        {
-            TC_LOG_ERROR("network", "%sSpeedChange player %s is NOT correct (must be %f instead %f), force set to correct value",
-                move_type_name[move_type], _player->GetName().c_str(), mover->GetSpeed(move_type), newspeed);
-            mover->SetSpeedRate(move_type, mover->GetSpeedRate(move_type));
-        }
-        else                                                // must be lesser - cheating
-        {
-            TC_LOG_DEBUG("misc", "Player %s from account id %u kicked for incorrect speed (must be %f instead %f)",
-                _player->GetName().c_str(), _player->GetSession()->GetAccountId(), mover->GetSpeed(move_type), newspeed);
-            _player->GetSession()->KickPlayer("WorldSession::HandleForceSpeedChangeAck Incorrect speed");
-        }
+        TC_LOG_WARN("misc", "The computed movement time using clockDelta is erronous. Using fallback instead");
+        movementInfo.time = GameTime::GetGameTimeMS();
+    }
+    else
+    {
+        movementInfo.time = (uint32)movementTime;
     }
+
+    mover->m_movementInfo = movementInfo;
+    mover->UpdatePosition(movementInfo.pos);
```

Comment:
just a doubt. In case of splines sent by movementgenerator or manually, there will be two UpdatePositions right? one done by the updatespline method and one on movement ack here. right?

Reply:
Sorry, i'm not sure I follow you. if you could give more context to your question, that would be appreciated

Reply:
While a unit is being moved by a spline (player charge or knockbacks) the spline system does its own position relocation in the Unit::Update loop. So what he probably asks for is disabling movementInfo based position relocations while the spline is doing its job and use the spline's position instead.

Reply:
not trying to imply any solution myself though, I was just stating the scenario. Don't know what's more appropiate, two updatepositions, the one on updatespline or the one on ack.

Reply:
IMO we should give the client as little control over such important value as possible so while we do have the chance of using spline data, we should do so let alone to avoid packet spoofing possibilities.

Reply:
These 2 lines are in no way different from the lines https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Handlers/MovementHandler.cpp#L396 & https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Handlers/MovementHandler.cpp#L415 in the `HandleMovementOpcodes` handler. The client is allowed to submit positional changes for the units it controls.  The ACK messages carry completely valid positions. The unit did went through these locations (at some point in the past). Disregarding them can actually lead to jerky movements when observing other players.

> the spline system does its own position relocation in the Unit::Update loop

For the majority of the time, spline will only run on units controlled by the server. And here, i'm talking about "real-time client control" (see [this PR](https://github.com/TrinityCore/TrinityCore/pull/26348)). The client is not allowed to send movement messages about them (if it does, they are ignored). So most of the time, there won't be a conflict. Player charges are indeed an exception. But even in retail, charges can lead to jerky movements. It is how the movement system works in WoW. Clients are authoritative on movement for the units under their controls.

PS: Knockback doesn't require a spline (the server can send SMSG_MOVE_KNOCK_BACK to the client, then the physical simulation of the knockback is run on the client).